tsdb/meta.go

```diff
@@ -927,18 +927,68 @@ func (m *Measurement) idsForExpr(n *influxql.BinaryExpr) (SeriesIDs, influxql.Ex
 				// return series that have a tag of specific value.
 				ids = tagVals[str.Val]
 			} else {
-				ids = m.seriesIDs
-				for k := range tagVals {
-					ids = ids.Reject(tagVals[k])
+				// Make a copy of all series ids and mark the ones we need to evict.
+				type id struct {
+					val   uint64
+					evict bool
+				}
+				seriesIDs := make([]id, len(m.seriesIDs))
+				for i, id := range m.seriesIDs {
+					seriesIDs[i].val = id
+				}
+				sz := len(seriesIDs)
+
+				// Go through each slice and mark the values we find as zero so
+				// they can be removed later.
+				for _, a := range tagVals {
+					tmp := seriesIDs
+					for _, id := range a {
+						if len(tmp) == 0 {
+							break
+						}
+
+						// Perform a binary search of the remaining slice if
+						// the first element does not match the value we're
+						// looking for.
+						i := 0
+						if tmp[0].val < id {
+							i = sort.Search(len(tmp), func(i int) bool {
+								return tmp[i].val >= id
+							})
+						}
+
+						if i >= len(tmp) {
+							break
+						} else if tmp[i].val == id {
+							if !tmp[i].evict {
+								tmp[i].evict = true
+								sz--
+							}
+							i++
+						}
+						tmp = tmp[i:]
+					}
+				}
+
+				// Make a new slice with only the remaining ids.
+				if sz > 0 {
+					ids = make([]uint64, 0, sz)
+					for _, id := range seriesIDs {
+						if id.evict {
+							continue
+						}
+						ids = append(ids, id.val)
+					}
 				}
```

Comment:
I think this all makes sense but can you refactor it to a function so it's more isolated and doesn't indent out so much. It makes it difficult to follow.